core/lib/node/src/resource/mod.rs

```diff
@@ -0,0 +1,35 @@
+use std::{any::Any, fmt};
+
+pub mod object_store;
+pub mod pools;
+
+/// A marker trait for anything that can be stored (and retrieved) as a resource.
+/// Requires `Clone` since the same resource may be requested by several tasks.
+pub trait Resource: 'static + Clone + std::any::Any {
+    /// Unique identifier of the resource.
+    /// Used to fetch the resource from the provider.
+    ///
+    /// It is recommended to name resources in form of `<scope>/<name>`, where `<scope>` is the name of the task
+    /// that will use this resource, or 'common' in case it is used by several tasks, and `<name>` is the name
+    /// of the resource itself.
+    const RESOURCE_NAME: &'static str;
+}
+
+/// An entity that knows how to initialize resources.
+///
+/// It exists to simplify the initialization process, as both tasks and *resources* can depend on other resources,
+/// and by having an entity that can initialize the resource on demand we can avoid the need to provide resources
```

Comment:
but resources cannot depend on tasks, so unless we want/need laziness, we could as well construct all the resources before initializing ZkSyncNode, therefore moving the dependency resolution to compilation time (i.e. profit).

Reply:
The problem here is that, for example, we can run the server with or without `eth_sender`, and depending on that, the operator's private key (and other configs in general) may not be available. So we kind of need laziness.

Reply:
ok, I assumed that the feasible execution modes will be hardcoded (like external node/main node). Now I understand that you want the set of tasks to be defined in an external config, instead of writing a short rust program which composes the required components. Is that correct?

Reply:
Yup!

>instead of writing a short rust program which composes the required components. Is that correct?

Yes, though it's just a hypothesis we want to test, not a commitment for X years of LTS. We will ship it (as it's relatively cheap to try) and see how it's being used -- if people would use framework more than external configs, we can always change things to favor that.